File: swap.clj
;; swap! already exists in Clojure
(defmacro my-swap! [x y]
  `(dosync
    (let [tmp# (deref ~x)]
      (ref-set ~x (deref ~y))
      (ref-set ~y tmp#))))

(def a (ref 1))
(def b (ref 2))

(my-swap! a b)
(prn "a")
(prn @a)
(prn "b")
(prn @b)

File: swap.rs
#![feature(macro_rules)]

macro_rules! swap {
    ($x:expr, $y:expr) => {
        {
            let tmp = $x;
            $x = $y;
            $y = tmp;
        }
    }
}

fn main() {
    let mut a = 1i;
    let mut b = 2i;

    swap!(a, b);
    println!("a: {} b: {}", a, b);

    let mut c = [3i, 4i];

    swap!(c[0], c[1]);
    println!("c[0]: {} c[1]: {}", c[0], c[1]);
}
